includes/modules/pages/index/header_php.php

```diff
@@ -13,40 +13,118 @@
+               FROM " . TABLE_CATEGORIES . "
+              WHERE parent_id = :currentCategoryId
+              LIMIT 1";
+        $category_parent_query = $db->bindVars($category_parent_query, ':currentCategoryId', $current_category_id, 'integer');
+        $category_parent = $db->Execute($category_parent_query);
+        $current_category_has_subcats = !$category_parent->EOF;
+    }
+    
+    // -----
+    // Give an observer the chance to override the default handling for the category.
+    //
+    $category_redirect_handled = false;
+    $zco_notifier->notify(
+        'NOTIFY_HEADER_INDEX_HTTP_STATUS_CHECK', 
+        array('cPath' => $cPath, 'current_category_id' => $current_category_id),
+        $category_redirect_handled,
+        $current_category_not_found,
+        $current_category_is_disabled,
+        $current_category_has_products,
+        $current_category_has_subcats,
+        $category_depth
+    );
+    
+    // -----
+    // If an observer hasn't overridden the default handling for the category's display:
+    //
+    // 1. Make sure that the current category-id is found for the store.  If not:
+    //    - Remove the 'cPath' parameter for follow-on processing by other modules.
+    //    - Reset the breadcrumbs.
+    //    - Set the flag to cause noindex/nofollow to be included for the page
+    //    - Issue a 404 (Not found)
+    // 2. Otherwise, make sure that the current category-id is enabled.  If not:
+    //    - Set the category_depth to indicate that a products' listing is to be displayed; it'll indicate no products found.
+    //    - Set the flag to cause noindex/nofollow to be included for the page
+    //    - Issue a 410 (Gone).
+    //
+    //    Note: A pseudo-configuration setting is used to allow stores to operate as in Zen Cart versions
+    //    prior to v157a, where a disabled category is still displayed.
+    //
+    // 3. Otherwise, the category is present and not disabled. Determine the category 'depth' to be displayed.
+    //    a. If the current category contains at least one product, display a products' listing.
+    //    b. Otherwise, check to see if the current category has sub-categories.  If so,
+    //       display a categories' listing; otherwise, display a products' listing, enabling the
+    //      'No products in category' message to be displayed.
+    //
+    //      Note: While this final check appears to be redundant, it's maintaining backward compatibility
+    //      of display for stores that have an 'invalid' mix of products and categories within a
+    //      category.
+    //
+    if (!$category_redirect_handled) {
+        if ($current_category_not_found) {
+            unset($_GET['cPath']);
+            $breadcrumb->reset();
+            $robotsNoIndex = true;
+            header('HTTP/1.1 404 Not Found');
+        } elseif ($current_category_is_disabled && (!defined('DISPLAY_DISABLED_CATEGORIES') || DISPLAY_DISABLED_CATEGORIES != 'true')) {
```

Comment:
Do you have any strong objection to removing this "pseudo-configuration setting"?

I'd prefer that people just use an observer if they want this sort of thing, particularly since it also provides infinitely more flexibility.

(I realize it was inspired by a similar setting for products .. which I'm very close to removing for similar reasons, particularly since it's not actually even working right now.)

Reply:
Depends on what you consider `strong`.  It seems like having some quick-and-easy way for stores to revert to previous behavior is a good thing.

Reply:
I'm not sure much fallback-to-old is necessary here.
Here's my reasoning:
- in either case it's going to cascade into `$category_depth = 'products'` as before
- the `$robotsNoIndex` isn't a destructive thing, so no harm in it being triggered without writing code to force old approach
- the `410` response is indeed the correct response for the situation, and will fine-tune SERP listings by the cleanup it offers, resulting in a better customer shopping experience. The notifier allows store-specific specialized handling if something more sophisticated is required.

Reply:
I don't quite understand.  Without that final bit, you'd never get to a `nested` sub-category listing.

Reply:
> you'd never get to a nested sub-category listing

Argh. I guess. 

I really don't like adding more constants. At least this one's not another admin configuration switch. 

I also don't like giving constants a string value of `'true'` or `'false'` when those words should only be booleans. (Notwithstanding the horrible precent in existing code, inherited from osC.) Better would be `'yes'` or `'no'` because there's no ambiguity or accidental misuse.
Better still is not to implement them at all. 

> It seems like having some quick-and-easy way for stores to revert to previous behavior is a good thing.

I'm fully compassionate to storeowners' resistance to change, and their varied "requirements"

... but do you think there's a very strong likelihood that droves of stores will scream for previous behavior?
If so, why are we even changing things?

Trying here to avoid adding technical debt of code that can never be removed for fear of breaking something one random store might be using.

I'd kinda rather tell the few people who need to "revert to previous" how to comment-out 3 lines of code for their specific use-case, instead of offering a switch that we're obligated to maintain for forever.

Reply:
Done, review requested.